radio/src/telemetry/flysky_ibus.cpp

```diff
@@ -0,0 +1,176 @@
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ */
+
+#include "opentx.h"
+#include "flysky_ibus.h"
+#include "../dataconstants.h"
+
+/*
+ * Telemetry format from goebish/Deviation/flysky_afhds2a_a7105.c
+ *
+ * format from RX:
+ * AA | TXID | RXID | sensor id | sensor # | value 16 bit big endian | sensor id ......
+ *  max 7 sensors per packet
+ *
+ *  TXID + RXID are already skipped in MULTI module to save memory+transmission time, format from Multi is:
+ *  AA | TX_RSSI | sensor ...
+ *
+ * OpenTX Mapping
+ *
+ *  instance = sensor id
+ *
+ *  Additional sensors from https://github.com/cleanflight/cleanflight/blob/master/src/main/telemetry/ibus.c
+ *
+ */
+
+
+#define FLYSKY_TELEMETRY_LENGTH (2+7*4)
+#define TX_RSSI_ID              300 // Pseudo id outside 1 byte range of FlySky sensors
+
+
+struct FlySkySensor {
+  const uint16_t id;
+  const char *name;
+  const TelemetryUnit unit;
+  const uint8_t precision;
+};
+
+
+const FlySkySensor flySkySensors[] = {
+
+  // RX Voltage (remapped, really 0x0)
+  {0x100,           ZSTR_A1,                UNIT_VOLTS,                  2},
+  // Temperature
+  {0x01,            ZSTR_TEMP1,             UNIT_CELSIUS,                1},
+  // RPM
+  {0x02,            ZSTR_RPM,               UNIT_RAW,                    0},
+  // External voltage
+  {0x03,            ZSTR_A3,                UNIT_VOLTS,                  1},
+  // RX Noise
+  {0xfb,            ZSTR_RX_SNR,            UNIT_DB,                     0},
+  // RX RSSI
+  {0xfc,            ZSTR_RSSI,              UNIT_DB,                     0},
+  // RX error rate
+  {0xfe,            ZSTR_RX_QUALITY,        UNIT_RAW,                    1},
+  // 0xff is an used sensor slot)
+  // Pseud sensor for TRSSI
+  {TX_RSSI_ID,      ZSTR_TX_RSSI,           UNIT_RAW,                     0},
+  // sentinel
+  {0x00,            NULL,                   UNIT_RAW,                    0},
+
+
+};
+
+static void processFlySkySensor(const uint8_t *packet) {
+  uint16_t id = packet[0];
+  const uint8_t instance = packet[1];
+  const int32_t value = (packet[3] << 8)  + packet[2];
+
+  if (id == 0xff)
+    // No sensor
+    return;
+  if (id == 0)
+    // Some part of OpenTX does not like sensor with id and instance 0, remap to 0x100
+    id = 0x100;
+
+  if(id == 0xfe)
+    telemetryData.rssi.set(value);
+
+  for (const FlySkySensor * sensor = flySkySensors; sensor->id; sensor++) {
+    // Extract value, skip header
+
+    setTelemetryValue(TELEM_PROTO_FLYSKY_IBUS, id, 0, instance, value, sensor->unit, sensor->precision);
+    return;
```

Comment:
Looked into it, it is loop until we find a sensor and otherwise set unknown sensor. All telemetry protocols do that :)